zetaclient/chains/evm/signer/outbound_data.go

```diff
@@ -45,143 +46,159 @@ type OutboundData struct {
-	cctxIndex, err := hex.DecodeString(cctx.Index[2:]) // remove the leading 0x
-	if err != nil || len(cctxIndex) != 32 {
-		return nil, true, fmt.Errorf("decode CCTX %s error", cctx.Index)
+	// Determine gas fees
+	gas, err := makeGasFromCCTX(cctx, evmObserver.GetPriorityGasFee(), logger)
+	if err != nil {
+		return nil, false, errors.Wrap(err, "failed to determine gas fees")
 	}
-	copy(txData.cctxIndex[:32], cctxIndex[:32])
 
 	// In case there is a pending transaction, make sure this keysign is a transaction replacement
-	pendingTx := evmObserver.GetPendingTx(nonce)
-	if pendingTx != nil {
-		if txData.gasPrice.Cmp(pendingTx.GasPrice()) > 0 {
-			logger.Info().
-				Msgf("replace pending outbound %s nonce %d using gas price %d", pendingTx.Hash().Hex(), nonce, txData.gasPrice)
-		} else {
-			logger.Info().Msgf("please wait for pending outbound %s nonce %d to be included", pendingTx.Hash().Hex(), nonce)
+	if tx := evmObserver.GetPendingTx(nonce); tx != nil {
+		newFeeIsLessThanPending := gas.MaxFeePerUnit.Cmp(tx.GasPrice()) <= 0
+
+		evt := logger.Info().
+			Str("cctx.pending_tx_hash", tx.Hash().Hex()).
+			Uint64("cctx.pending_tx_nonce", nonce)
+
+		if newFeeIsLessThanPending {
+			evt.Msg("Please wait for pending outbound to be included in the block")
 			return nil, true, nil
 		}
+
+		evt.
+			Uint64("cctx.max_gas_fee", gas.MaxFeePerUnit.Uint64()).
+			Msg("Replacing pending outbound transaction with higher gas fees")
 	}
 
 	// Base64 decode message
+	var message []byte
 	if cctx.InboundParams.CoinType != coin.CoinType_Cmd {
-		txData.message, err = base64.StdEncoding.DecodeString(cctx.RelayedMessage)
-		if err != nil {
-			logger.Err(err).Msgf("decode CCTX.Message %s error", cctx.RelayedMessage)
+		msg, errDecode := base64.StdEncoding.DecodeString(cctx.RelayedMessage)
+		if errDecode != nil {
+			logger.Err(err).Str("cctx.relayed_message", cctx.RelayedMessage).Msg("Unable to decode relayed message")
+		} else {
+			message = msg
 		}
 	}
 
-	return &txData, false, nil
+	logger.Info().
+		Str("cctx.to_chain_name", toChain.GetChainName().String()).
+		Int64("cctx.to_chain_id", toChain.ChainId).
+		Str("cctx.to_recipient", to.Hex()).
+		Uint64("cctx.nonce", nonce).
+		Uint64("cctx.finalized_zeta_height", cctx.InboundParams.FinalizedZetaHeight).
+		Msg("Constructed OutboundData")
+
+	return &OutboundData{
+		outboundParams: outboundParams,
+
+		srcChainID: big.NewInt(cctx.InboundParams.SenderChainId),
+		sender:     ethcommon.HexToAddress(cctx.InboundParams.Sender),
+
+		toChainID: toChainID,
+		to:        to,
+
+		asset:  ethcommon.HexToAddress(cctx.InboundParams.Asset),
+		amount: outboundParams.Amount.BigInt(),
+
+		gas:    gas,
+		nonce:  outboundParams.TssNonce,
+		height: height,
+
+		message: message,
+
+		cctxIndex: cctxIndex,
+	}, false, nil
+}
+
+func getCCTXIndex(cctx *types.CrossChainTx) ([32]byte, error) {
+	cctxIndexSlice, err := hex.DecodeString(cctx.Index[2:]) // remove the leading 0x
```

Comment:
This would probably panic (depending how DecodeString is implemented) if cctx is null. Consider adding a null check before this.